File: Tests/TrainingSafeguardTests.swift
// SPDX-License-Identifier: MPL-2.0
// Original contributions Copyright 2026 Nuri Bruner.

import XCTest
@testable import Doigt

final class TrainingSafeguardTests: XCTestCase {
    func testImportedMaxTableRejectsNonFiniteAndNonPositiveValues() {
        for invalid in [Double.infinity, -.infinity, .nan, 0, -5] {
            let table = MaxTable(keyed: [MaxTable.key(grip: GripSpec().key, side: .both): invalid])
            XCTAssertNil(table.max(grip: GripSpec().key, side: .left))
        }
        let table = MaxTable(keyed: [MaxTable.key(grip: GripSpec().key, side: .both): 40])
        XCTAssertEqual(table.max(grip: GripSpec().key, side: .left), 40)
    }

    func testLaterPercentageSetWithSameGripStillReportsMissingBenchmark() {
        var plan = SessionPlan()
        plan.sets = [SetPlan(targetLoKg: 5), SetPlan(targetLoPercent: 0.25), SetPlan(targetLoPercent: 0.3)]
        XCTAssertEqual(PlanMath.untargetedGripCount(plan, maxes: MaxTable()), 1)
        XCTAssertEqual(PlanMath.missingBenchmarkGripCount(plan, maxes: MaxTable()), 1)
    }

    func testFixedTargetsUnplannedTargetsAndSkippedSetsDoNotRequireBenchmark() {
        var plan = SessionPlan()
        plan.sets = [SetPlan(targetLoKg: 5), SetPlan(), SetPlan(repsPerSide: 0, targetLoPercent: 0.25)]
        XCTAssertEqual(PlanMath.missingBenchmarkGripCount(plan, maxes: MaxTable()), 0)
    }

    func testPercentageTargetsNeedBothHandsBeforeWarningDisappears() {
        var plan = RoutineDraft.starter.plan
        plan.sets = [SetPlan(targetLoPercent: 0.25)]
        var maxes = MaxTable()
        maxes.record(40, grip: GripSpec().key, side: .left)
        XCTAssertEqual(PlanMath.missingBenchmarkGripCount(plan, maxes: maxes), 1)
        maxes.record(50, grip: GripSpec().key, side: .right)
        XCTAssertEqual(PlanMath.missingBenchmarkGripCount(plan, maxes: maxes), 0)
    }
}

File: android/engine/src/test/kotlin/run/nuri/getagrip/engine/TrainingSafeguardTests.kt
// SPDX-License-Identifier: MPL-2.0
// Original contributions Copyright 2026 Nuri Bruner.

package run.nuri.getagrip.engine

import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertNull

class TrainingSafeguardTests {
    @Test fun importedMaxTableRejectsNonFiniteAndNonPositiveValues() {
        for (invalid in listOf(Double.POSITIVE_INFINITY, Double.NEGATIVE_INFINITY, Double.NaN, 0.0, -5.0)) {
            val table = MaxTable(mapOf(MaxTable.key(GripSpec().key, Side.both) to invalid))
            assertNull(table.max(GripSpec().key, Side.left))
        }
        val table = MaxTable(mapOf(MaxTable.key(GripSpec().key, Side.both) to 40.0))
        assertEquals(40.0, table.max(GripSpec().key, Side.left))
    }

    @Test fun laterPercentageSetWithSameGripStillReportsMissingBenchmark() {
        val plan = SessionPlan(sets = listOf(SetPlan(targetLoKg = 5.0), SetPlan(targetLoPercent = 0.25), SetPlan(targetLoPercent = 0.3)))
        assertEquals(1, PlanMath.untargetedGripCount(plan, MaxTable()))
        assertEquals(1, PlanMath.missingBenchmarkGripCount(plan, MaxTable()))
    }

    @Test fun fixedTargetsUnplannedTargetsAndSkippedSetsDoNotRequireBenchmark() {
        val plan = SessionPlan(sets = listOf(SetPlan(targetLoKg = 5.0), SetPlan(), SetPlan(repsPerSide = 0, targetLoPercent = 0.25)))
        assertEquals(0, PlanMath.missingBenchmarkGripCount(plan, MaxTable()))
    }

    @Test fun percentageTargetsNeedBothHandsBeforeWarningDisappears() {
        val plan = RoutineDraft.starter.plan.copy(sets = listOf(SetPlan(targetLoPercent = 0.25)))
        val maxes = MaxTable()
        maxes.record(40.0, GripSpec().key, Side.left)
        assertEquals(1, PlanMath.missingBenchmarkGripCount(plan, maxes))
        maxes.record(50.0, GripSpec().key, Side.right)
        assertEquals(0, PlanMath.missingBenchmarkGripCount(plan, maxes))
    }
}
